# Algoritmo: Subida de Encosta
# Problema: O caixeiro Viajante

# 1. Código

In [2]:
# Biblioteca auxiliar
import pandas as pd

In [1]:
import random
import math


# Função para gerar um vizinho trocando dois pontos na rota
def operador1(rota, i, j):
    vizinho = rota[:]  # cópia da rota
    vizinho[i], vizinho[j] = (
        vizinho[j],
        vizinho[i],
    )  # troca a cidade que estava no índice i para o índice j, e vice-versa
    return vizinho


# Função para inverter trechos das permutações entre dois pontos da rota
def operador2(rota, i, j):
    tam = len(rota)  # tamanho da rota
    vizinho = rota[:]  # cópia da rota
    if i > j:  # caso i seja maior que j (propriedade circular da lista)
        for c in vizinho[
            : j - i
        ]:  # aumenta a lista para auxiliar na inversão de trechos
            vizinho.append(c)
        while j >= 0:  # realiza a troca de cidades (garante a inversão)
            vizinho[j], vizinho[i] = vizinho[i], vizinho[j]
            i += 1
            j -= 1
        vizinho = vizinho[:tam]  # transforma de volta a lista para seu tamanho original
    else:
        while i < j:  # caso i menor que j (trecho interno da rota)
            vizinho[i], vizinho[j] = (
                vizinho[j],
                vizinho[i],
            )  # faz troca de cidades (garante a inversão)
            i += 1
            j -= 1
            if i == tam:  # garante índices dentro do intervalo
                i == 0
            if j < 0:  # garante índices dentro do intervalo
                j = tam - 1
    return vizinho


# Função para calcular a distância entre duas cidades
def distancia(cidade1, cidade2):
    return math.sqrt((cidade1[0] - cidade2[0]) ** 2 + (cidade1[1] - cidade2[1]) ** 2)


# Função para calcular o custo total de uma rota
def calcular_custo(rota, pesos):
    custo = 0.0
    # percorre toda a rota somando os custos de uma cidade a outra
    # ainda calcula o custo da última cidade para a cidade de origem
    for i in range(len(rota)):
        cidade_atual = rota[i]
        proxima_cidade = rota[
            (i + 1) % len(rota)
        ]  # garante que o índice não saia do invervalo
        custo += pesos[cidade_atual][proxima_cidade]
    return custo


# Função para gerar vizinhos com base em uma rota
def calcular_vizinhos(cidades, rota, operador):
    vizinhos = []  # lista de vizinhos
    for i in range(
        len(cidades)
    ):  # para cada cidade, calcula novos vizinhos com base na rota e no operador escolhido
        for j in range(len(cidades)):
            vizinhos.append(operador(rota, i, j))
    return vizinhos


# Subida de Encosta
def subida_de_encosta(
    cidades, operador, pesos, inicialRandom=True, randomizar=True
):

    inicial = list(range(len(cidades)))  # estado inicial é a sequencia lida no arquivo
    if (
        inicialRandom
    ):  # caso o estado inicial seja gerado aleatoriamente, embaralhamos a sequencia lida no arquivo
        random.shuffle(inicial)
    melhor_rota = inicial[:]  # melhor rota atual é a inicial
    melhor_custo = calcular_custo(inicial, pesos)  # melhor custo atual

    condicao = True # condição de parada do loop
    # executa até não haver um vizinho com custo melhor
    while condicao:
        condicao = False # Até agora nenhum vizinho foi melhor
        vizinhos = calcular_vizinhos(cidades, melhor_rota, operador) # encontra os vizinhos possíveis

        if randomizar: # se a variação for para randomizar os vizinhos, randomiza
            random.shuffle(vizinhos)
        # ver o primeiro vizinho com custo menor
        for vizinho in vizinhos:
            custo_vizinho = calcular_custo(vizinho, pesos) # calcula o custo do vizinho
            if custo_vizinho < melhor_custo: # verifica se o custo do vizinho é menor que o melhor custo
                melhor_rota = vizinho # atualiza o valor da melhor rota
                melhor_custo = custo_vizinho # atualiza o valor do melhor custo
                condicao = True # deixa True para verificar os próximos vizinhos
                break

    return melhor_custo, melhor_rota

- Organização de Cidades e Cálculo de pesos

In [12]:
# Função para ler as coordenadas das cidades
def leitura(entrada_x, entrada_y):
    X = [float(x) for x in entrada_x]  # guarda os valores de x
    Y = [float(y) for y in entrada_y]   # guarda os valores de y
    cidades = [(X[i], Y[i]) for i in range(len(X))]  # organiza as cidades
    pesos = []
    # calcula os pesos (a distância) de uma cidade para outra
    for cidade1 in cidades:
        coluna = []
        for cidade2 in cidades:
            coluna.append(distancia(cidade1, cidade2))
        pesos.append(coluna)
    # retorna as coordenadas das cidades e os pesos de cada cidade
    return cidades, pesos

# 2. Bloco de Experimentação

## 2.1. Entradas: Coordenadas das cidades

In [14]:
# Valores de x das cidades
entrada_x = ['0.0', '1.0', '2.0', '0.0', '1.0', '1.0', '3.0']
# Valores de y das cidades
entrada_y = ['0.0', '2.0', '0.0', '2.0', '1.0', '3.0', '2.0']

## 2.2. Testes para cada variação

- Estado Inicial 1: segue a ordem de leitura
- Estado Inicial 2: arbitrário

In [21]:
def testes(cidades, operador, pesos, inicialRandom, randomizar):
  df = pd.DataFrame()
  lista = []
  for _ in range(30):
    melhor_custo, melhor_rota = subida_de_encosta(cidades, operador, pesos, inicialRandom, randomizar)
    lista.append((melhor_custo, melhor_rota))
  df = pd.DataFrame(lista, columns=[["Melhor Custo", "Melhor Rota"]])
  return df

- Leitura de cidades e pesos

In [15]:
cidades, pesos = leitura(entrada_x, entrada_y)

### 2.2.1. Variação 1: Estado Inicial 1 com Operador 1 sem randomização da vizinhança

In [45]:
# Estado Inicial 1 com Operador 1 sem randomização da vizinhança
df1 = testes(cidades, operador1, pesos, False, False)
display(df1)

Melhor Custo            Melhor Rota
0     11.300563  [5, 1, 3, 4, 0, 2, 6]
1     11.300563  [5, 1, 3, 4, 0, 2, 6]
2     11.300563  [5, 1, 3, 4, 0, 2, 6]
3     11.300563  [5, 1, 3, 4, 0, 2, 6]
4     11.300563  [5, 1, 3, 4, 0, 2, 6]
5     11.300563  [5, 1, 3, 4, 0, 2, 6]
6     11.300563  [5, 1, 3, 4, 0, 2, 6]
7     11.300563  [5, 1, 3, 4, 0, 2, 6]
8     11.300563  [5, 1, 3, 4, 0, 2, 6]
9     11.300563  [5, 1, 3, 4, 0, 2, 6]
10    11.300563  [5, 1, 3, 4, 0, 2, 6]
11    11.300563  [5, 1, 3, 4, 0, 2, 6]
12    11.300563  [5, 1, 3, 4, 0, 2, 6]
13    11.300563  [5, 1, 3, 4, 0, 2, 6]
14    11.300563  [5, 1, 3, 4, 0, 2, 6]
15    11.300563  [5, 1, 3, 4, 0, 2, 6]
16    11.300563  [5, 1, 3, 4, 0, 2, 6]
17    11.300563  [5, 1, 3, 4, 0, 2, 6]
18    11.300563  [5, 1, 3, 4, 0, 2, 6]
19    11.300563  [5, 1, 3, 4, 0, 2, 6]
20    11.300563  [5, 1, 3, 4, 0, 2, 6]
21    11.300563  [5, 1, 3, 4, 0, 2, 6]
22    11.300563  [5, 1, 3, 4, 0, 2, 6]
23    11.300563  [5, 1, 3, 4, 0, 2, 6]
24    11.300563  [5, 1, 3, 4, 0, 2, 6]
25    11.300563  [5, 1, 3, 4, 0, 2, 6]
26    11.300563  [5, 1, 3, 4, 0, 2, 6]
27    11.300563  [5, 1, 3, 4, 0, 2, 6]
28    11.300563  [5, 1, 3, 4, 0, 2, 6]
29    11.300563  [5, 1, 3, 4, 0, 2, 6]

### 2.2.2. Variação 2: Estado Inicial 1 com Operador 1 com randomização da vizinhança

In [24]:
# Estado Inicial 1 com Operador 1 com randomização da vizinhança
df2 = testes(cidades, operador1, pesos, False, True)
display(df2)

Melhor Custo            Melhor Rota
0     11.300563  [5, 6, 2, 0, 4, 3, 1]
1     11.300563  [0, 4, 2, 6, 5, 1, 3]
2     11.300563  [5, 6, 2, 0, 4, 3, 1]
3     11.300563  [0, 2, 6, 5, 3, 1, 4]
4     11.886350  [4, 1, 5, 3, 0, 2, 6]
5     11.886350  [1, 4, 6, 2, 0, 3, 5]
6     12.064495  [2, 0, 3, 5, 6, 1, 4]
7     12.064495  [0, 2, 4, 1, 6, 5, 3]
8     11.886350  [4, 6, 2, 0, 3, 5, 1]
9     11.300563  [2, 6, 5, 1, 3, 4, 0]
10    11.300563  [3, 1, 5, 6, 2, 4, 0]
11    11.300563  [6, 2, 0, 4, 1, 3, 5]
12    11.300563  [6, 5, 3, 1, 4, 0, 2]
13    11.300563  [5, 1, 3, 4, 0, 2, 6]
14    12.064495  [5, 3, 0, 2, 4, 1, 6]
15    11.886350  [0, 3, 5, 1, 4, 6, 2]
16    11.886350  [4, 1, 5, 3, 0, 2, 6]
17    11.886350  [3, 0, 2, 6, 4, 1, 5]
18    11.300563  [5, 1, 3, 0, 4, 2, 6]
19    11.300563  [2, 0, 4, 3, 1, 5, 6]
20    11.300563  [2, 4, 0, 3, 1, 5, 6]
21    11.300563  [1, 3, 0, 4, 2, 6, 5]
22    11.300563  [6, 2, 0, 4, 3, 1, 5]
23    11.300563  [3, 0, 4, 2, 6, 5, 1]
24    11.300563  [3, 4, 0, 2, 6, 5, 1]
25    11.300563  [5, 6, 2, 0, 4, 3, 1]
26    11.300563  [5, 6, 2, 4, 0, 3, 1]
27    11.300563  [1, 5, 6, 2, 4, 0, 3]
28    11.300563  [3, 0, 4, 2, 6, 5, 1]
29    11.300563  [1, 5, 6, 2, 4, 0, 3]

### 2.2.3. Variação 3: Estado Inicial 1 com Operador 2 sem randomização da vizinhança

In [27]:
# Estado Inicial 1 com Operador 2 sem randomização da vizinhança
df3 = testes(cidades, operador2, pesos, False, False)
display(df3)

Melhor Custo            Melhor Rota
0     11.300563  [2, 0, 4, 1, 3, 5, 6]
1     11.300563  [2, 0, 4, 1, 3, 5, 6]
2     11.300563  [2, 0, 4, 1, 3, 5, 6]
3     11.300563  [2, 0, 4, 1, 3, 5, 6]
4     11.300563  [2, 0, 4, 1, 3, 5, 6]
5     11.300563  [2, 0, 4, 1, 3, 5, 6]
6     11.300563  [2, 0, 4, 1, 3, 5, 6]
7     11.300563  [2, 0, 4, 1, 3, 5, 6]
8     11.300563  [2, 0, 4, 1, 3, 5, 6]
9     11.300563  [2, 0, 4, 1, 3, 5, 6]
10    11.300563  [2, 0, 4, 1, 3, 5, 6]
11    11.300563  [2, 0, 4, 1, 3, 5, 6]
12    11.300563  [2, 0, 4, 1, 3, 5, 6]
13    11.300563  [2, 0, 4, 1, 3, 5, 6]
14    11.300563  [2, 0, 4, 1, 3, 5, 6]
15    11.300563  [2, 0, 4, 1, 3, 5, 6]
16    11.300563  [2, 0, 4, 1, 3, 5, 6]
17    11.300563  [2, 0, 4, 1, 3, 5, 6]
18    11.300563  [2, 0, 4, 1, 3, 5, 6]
19    11.300563  [2, 0, 4, 1, 3, 5, 6]
20    11.300563  [2, 0, 4, 1, 3, 5, 6]
21    11.300563  [2, 0, 4, 1, 3, 5, 6]
22    11.300563  [2, 0, 4, 1, 3, 5, 6]
23    11.300563  [2, 0, 4, 1, 3, 5, 6]
24    11.300563  [2, 0, 4, 1, 3, 5, 6]
25    11.300563  [2, 0, 4, 1, 3, 5, 6]
26    11.300563  [2, 0, 4, 1, 3, 5, 6]
27    11.300563  [2, 0, 4, 1, 3, 5, 6]
28    11.300563  [2, 0, 4, 1, 3, 5, 6]
29    11.300563  [2, 0, 4, 1, 3, 5, 6]

### 2.2.4. Variação 4: Estado Inicial 1 com Operador 2 com randomização da vizinhança

In [28]:
# Estado Inicial 1 com Operador 2 com randomização da vizinhança
df4 = testes(cidades, operador2, pesos, False, True)
display(df4)

Melhor Custo            Melhor Rota
0     11.300563  [3, 5, 6, 2, 0, 4, 1]
1     11.300563  [6, 2, 4, 0, 3, 1, 5]
2     12.064495  [6, 5, 3, 0, 2, 4, 1]
3     11.300563  [3, 5, 6, 2, 0, 4, 1]
4     12.064495  [5, 3, 0, 2, 4, 1, 6]
5     11.300563  [5, 6, 2, 0, 4, 3, 1]
6     11.300563  [6, 5, 3, 1, 4, 0, 2]
7     11.300563  [6, 5, 1, 3, 4, 0, 2]
8     11.300563  [1, 3, 4, 0, 2, 6, 5]
9     11.300563  [5, 1, 3, 4, 0, 2, 6]
10    11.300563  [3, 4, 0, 2, 6, 5, 1]
11    11.300563  [6, 2, 0, 4, 3, 1, 5]
12    11.300563  [0, 4, 1, 3, 5, 6, 2]
13    11.300563  [2, 4, 0, 3, 1, 5, 6]
14    11.300563  [2, 0, 4, 3, 1, 5, 6]
15    11.300563  [5, 1, 3, 0, 4, 2, 6]
16    11.300563  [6, 2, 0, 4, 1, 3, 5]
17    11.300563  [5, 3, 1, 4, 0, 2, 6]
18    11.300563  [5, 3, 1, 4, 0, 2, 6]
19    11.300563  [5, 3, 1, 4, 0, 2, 6]
20    11.300563  [2, 4, 0, 3, 1, 5, 6]
21    11.300563  [3, 5, 6, 2, 0, 4, 1]
22    11.300563  [0, 4, 2, 6, 5, 1, 3]
23    11.886350  [6, 4, 1, 5, 3, 0, 2]
24    12.064495  [2, 4, 1, 6, 5, 3, 0]
25    11.300563  [5, 6, 2, 4, 0, 3, 1]
26    11.300563  [2, 0, 4, 3, 1, 5, 6]
27    11.300563  [1, 3, 0, 4, 2, 6, 5]
28    11.886350  [4, 1, 5, 3, 0, 2, 6]
29    11.300563  [5, 1, 3, 0, 4, 2, 6]

### 2.2.5. Variação 5: Estado Inicial 2 com Operador 1 sem randomização da vizinhança

In [29]:
# Estado Inicial 2 com Operador 1 sem randomização da vizinhança
df5 = testes(cidades, operador1, pesos, True, False)
display(df5)

Melhor Custo            Melhor Rota
0     11.300563  [6, 5, 1, 3, 4, 0, 2]
1     11.300563  [3, 1, 5, 6, 2, 4, 0]
2     11.300563  [4, 1, 3, 5, 6, 2, 0]
3     11.300563  [0, 4, 3, 1, 5, 6, 2]
4     12.064495  [3, 5, 6, 1, 4, 2, 0]
5     11.300563  [1, 3, 4, 0, 2, 6, 5]
6     12.064495  [1, 6, 5, 3, 0, 2, 4]
7     11.300563  [3, 5, 6, 2, 0, 4, 1]
8     11.886350  [2, 0, 3, 5, 1, 4, 6]
9     11.300563  [1, 3, 0, 4, 2, 6, 5]
10    11.300563  [6, 5, 1, 3, 0, 4, 2]
11    11.300563  [6, 2, 4, 0, 3, 1, 5]
12    11.300563  [0, 4, 3, 1, 5, 6, 2]
13    11.300563  [0, 4, 2, 6, 5, 1, 3]
14    11.300563  [0, 4, 3, 1, 5, 6, 2]
15    11.300563  [3, 1, 5, 6, 2, 4, 0]
16    11.300563  [3, 5, 6, 2, 0, 4, 1]
17    11.300563  [2, 0, 4, 1, 3, 5, 6]
18    11.300563  [3, 1, 5, 6, 2, 4, 0]
19    11.300563  [1, 3, 0, 4, 2, 6, 5]
20    11.300563  [3, 5, 6, 2, 0, 4, 1]
21    11.300563  [5, 6, 2, 0, 4, 3, 1]
22    11.300563  [1, 5, 6, 2, 4, 0, 3]
23    11.300563  [6, 5, 1, 3, 0, 4, 2]
24    11.886350  [4, 6, 2, 0, 3, 5, 1]
25    11.300563  [0, 2, 6, 5, 1, 3, 4]
26    11.300563  [6, 2, 4, 0, 3, 1, 5]
27    11.300563  [3, 1, 5, 6, 2, 4, 0]
28    11.300563  [2, 0, 4, 1, 3, 5, 6]
29    11.300563  [6, 5, 3, 1, 4, 0, 2]

### 2.2.6. Variação 6: Estado Inicial 2 com Operador 1 com randomização da vizinhança

In [30]:
# Estado Inicial 2 com Operador 1 com randomização da vizinhança
df6 = testes(cidades, operador1, pesos, True, True)
display(df6)

Melhor Custo            Melhor Rota
0     11.300563  [0, 4, 2, 6, 5, 1, 3]
1     11.300563  [6, 2, 4, 0, 3, 1, 5]
2     11.886350  [1, 4, 6, 2, 0, 3, 5]
3     12.064495  [2, 0, 3, 5, 6, 1, 4]
4     11.300563  [5, 6, 2, 4, 0, 3, 1]
5     11.886350  [0, 3, 5, 1, 4, 6, 2]
6     11.886350  [0, 2, 6, 4, 1, 5, 3]
7     11.300563  [3, 4, 0, 2, 6, 5, 1]
8     11.300563  [0, 4, 1, 3, 5, 6, 2]
9     11.300563  [4, 3, 1, 5, 6, 2, 0]
10    11.300563  [6, 2, 0, 4, 3, 1, 5]
11    11.300563  [1, 3, 4, 0, 2, 6, 5]
12    11.300563  [6, 5, 3, 1, 4, 0, 2]
13    11.300563  [1, 3, 0, 4, 2, 6, 5]
14    11.300563  [0, 4, 2, 6, 5, 1, 3]
15    11.300563  [5, 1, 3, 4, 0, 2, 6]
16    12.064495  [0, 2, 4, 1, 6, 5, 3]
17    11.300563  [0, 2, 6, 5, 3, 1, 4]
18    11.300563  [6, 2, 0, 4, 3, 1, 5]
19    11.300563  [6, 2, 0, 4, 3, 1, 5]
20    11.300563  [6, 5, 3, 1, 4, 0, 2]
21    11.886350  [0, 2, 6, 4, 1, 5, 3]
22    11.886350  [2, 6, 4, 1, 5, 3, 0]
23    11.300563  [3, 5, 6, 2, 0, 4, 1]
24    12.064495  [0, 2, 4, 1, 6, 5, 3]
25    11.300563  [6, 5, 1, 3, 0, 4, 2]
26    11.300563  [0, 4, 3, 1, 5, 6, 2]
27    12.064495  [5, 6, 1, 4, 2, 0, 3]
28    11.300563  [3, 5, 6, 2, 0, 4, 1]
29    11.300563  [4, 3, 1, 5, 6, 2, 0]

### 2.2.7. Variação 7: Estado Inicial 2 com Operador 2 sem randomização da vizinhança

In [31]:
# Estado Inicial 2 com Operador 2 sem randomização da vizinhança
df7 = testes(cidades, operador2, pesos, True, False)
display(df7)

Melhor Custo            Melhor Rota
0     11.300563  [3, 0, 4, 2, 6, 5, 1]
1     11.300563  [3, 0, 4, 2, 6, 5, 1]
2     11.300563  [5, 6, 2, 0, 4, 3, 1]
3     11.300563  [5, 6, 2, 0, 4, 3, 1]
4     11.300563  [6, 5, 3, 1, 4, 0, 2]
5     11.300563  [3, 1, 5, 6, 2, 4, 0]
6     11.886350  [0, 3, 5, 1, 4, 6, 2]
7     12.064495  [2, 4, 1, 6, 5, 3, 0]
8     11.300563  [3, 5, 6, 2, 0, 4, 1]
9     11.300563  [3, 4, 0, 2, 6, 5, 1]
10    11.300563  [0, 4, 1, 3, 5, 6, 2]
11    11.300563  [6, 2, 0, 4, 1, 3, 5]
12    11.300563  [6, 2, 0, 4, 3, 1, 5]
13    11.300563  [5, 1, 3, 0, 4, 2, 6]
14    11.886350  [0, 2, 6, 4, 1, 5, 3]
15    11.300563  [0, 4, 2, 6, 5, 1, 3]
16    11.300563  [6, 5, 1, 3, 4, 0, 2]
17    12.064495  [5, 3, 0, 2, 4, 1, 6]
18    11.300563  [0, 4, 1, 3, 5, 6, 2]
19    11.300563  [0, 4, 2, 6, 5, 1, 3]
20    11.300563  [0, 4, 1, 3, 5, 6, 2]
21    11.300563  [2, 0, 4, 1, 3, 5, 6]
22    11.300563  [3, 4, 0, 2, 6, 5, 1]
23    11.300563  [6, 2, 0, 4, 3, 1, 5]
24    11.886350  [1, 5, 3, 0, 2, 6, 4]
25    11.886350  [1, 5, 3, 0, 2, 6, 4]
26    11.300563  [2, 0, 4, 3, 1, 5, 6]
27    11.886350  [0, 3, 5, 1, 4, 6, 2]
28    11.300563  [2, 6, 5, 3, 1, 4, 0]
29    11.300563  [0, 4, 1, 3, 5, 6, 2]

### 2.2.8. Variação 8: Estado Inicial 2 com Operador 2 com randomização da vizinhança

In [32]:
# Estado Inicial 2 com Operador 2 com randomização da vizinhança
df8 = testes(cidades, operador2, pesos, True, True)
display(df8)

Melhor Custo            Melhor Rota
0     11.300563  [3, 1, 5, 6, 2, 4, 0]
1     11.300563  [3, 0, 4, 2, 6, 5, 1]
2     12.064495  [5, 3, 0, 2, 4, 1, 6]
3     11.300563  [5, 1, 3, 4, 0, 2, 6]
4     11.300563  [3, 0, 4, 2, 6, 5, 1]
5     11.300563  [2, 6, 5, 3, 1, 4, 0]
6     11.300563  [2, 0, 4, 3, 1, 5, 6]
7     11.300563  [2, 0, 4, 3, 1, 5, 6]
8     11.300563  [4, 0, 2, 6, 5, 3, 1]
9     11.300563  [3, 1, 5, 6, 2, 4, 0]
10    11.300563  [6, 2, 4, 0, 3, 1, 5]
11    11.300563  [2, 0, 4, 3, 1, 5, 6]
12    11.300563  [5, 3, 1, 4, 0, 2, 6]
13    11.300563  [5, 6, 2, 4, 0, 3, 1]
14    11.300563  [5, 6, 2, 0, 4, 3, 1]
15    11.300563  [5, 6, 2, 0, 4, 3, 1]
16    11.300563  [1, 3, 4, 0, 2, 6, 5]
17    11.886350  [6, 4, 1, 5, 3, 0, 2]
18    11.300563  [5, 1, 3, 4, 0, 2, 6]
19    11.300563  [6, 5, 3, 1, 4, 0, 2]
20    11.300563  [2, 6, 5, 1, 3, 4, 0]
21    12.064495  [1, 4, 2, 0, 3, 5, 6]
22    11.300563  [0, 4, 1, 3, 5, 6, 2]
23    11.300563  [3, 0, 4, 2, 6, 5, 1]
24    11.300563  [6, 2, 4, 0, 3, 1, 5]
25    11.300563  [5, 3, 1, 4, 0, 2, 6]
26    12.064495  [5, 3, 0, 2, 4, 1, 6]
27    11.886350  [2, 0, 3, 5, 1, 4, 6]
28    11.300563  [5, 3, 1, 4, 0, 2, 6]
29    11.300563  [1, 3, 4, 0, 2, 6, 5]

# 3. Análise dos resultados

- Tabela com média dos melhores custos de cada variação

In [37]:
# Lista com as médias
lista = []
lista.append(df1['Melhor Custo'].mean())
lista.append(df2['Melhor Custo'].mean())
lista.append(df3['Melhor Custo'].mean())
lista.append(df4['Melhor Custo'].mean())
lista.append(df5['Melhor Custo'].mean())
lista.append(df6['Melhor Custo'].mean())
lista.append(df7['Melhor Custo'].mean())
lista.append(df8['Melhor Custo'].mean())
df = pd.DataFrame(lista)

In [39]:
df.transpose()

0          1          2          3          4  \
Melhor Custo  11.300563  11.494114  11.300563  11.416009  11.390544   

                      5          6          7  
Melhor Custo  11.500052  11.449123  11.416009

- Melhor Custo

In [43]:
df.min()

Melhor Custo    11.300563
dtype: float64

Com os resultados obtidos, as variações com melhor custo médio foi a variação 1 e a variação 3, com valores de 11.300563.
Importante levar em consideração que esse foi o resultado obtido para cada uma das 30 execuções do código, isto é, a média dos custos encontrados foi igual ao melhor custo encontrado em cada uma das execuções dessas duas variações.